File: SKA/puzzles/day1.go
package puzzles

import (
	"bufio"
	"log"
	"regexp"
	"strconv"
	"strings"
)

func Day1Task1() {
	input := openTextFile("inputs/aoc01.txt")
	defer input.Close()

	regexPattern := "[0-9]"
	regex, err := regexp.Compile(regexPattern)
	if err != nil {
		log.Fatal(err)
	}

	sum := 0

	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()
		digits := regex.FindAllString(line, -1)
		if len(digits) == 0 {
			continue
		} else if len(digits) == 1 {
			numberString := digits[0]
			numberString = strings.Repeat(numberString, 2)
			number := stringToNumber(numberString)
			sum += number
		} else {
			numberString := digits[0] + digits[len(digits)-1]
			number := stringToNumber(numberString)
			sum += number

		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	println(sum)
}

func Day1Task2() {
	input := openTextFile("inputs/aoc01.txt")
	defer input.Close()

	numberMap := map[string]int{
		"one":   1,
		"two":   2,
		"three": 3,
		"four":  4,
		"five":  5,
		"six":   6,
		"seven": 7,
		"eight": 8,
		"nine":  9,
	}

	literalNumberConverter := regexp.MustCompile(`(one|two|three|four|five|six|seven|eight|nine)`)
	digitFinder := regexp.MustCompile("[1-9]")
	sum := 0

	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		line := scanner.Text()
		replacedNumbersString := literalNumberConverter.ReplaceAllStringFunc(line, func(matched string) string {
			return strconv.Itoa(numberMap[matched])
		})
		digits := digitFinder.FindAllString(replacedNumbersString, -1)

		if len(digits) == 0 {
			println("No digits found")
			continue
		} else if len(digits) == 1 {
			numberString := digits[0]
			numberString = strings.Repeat(numberString, 2)
			number := stringToNumber(numberString)
			sum += number
		} else {
			numberString := digits[0] + digits[len(digits)-1]
			number := stringToNumber(numberString)
			sum += number

		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	println(sum)

}

File: SKA/day1.py
"""Could not solve day 1 part 2 with my preferred solution because Go Regex does not support
positive lookahead. Python to the rescue."""

import re
from pathlib import PosixPath


def main():
    sum = 0
    regex = "1|2|3|4|5|6|7|8|9|one|two|three|four|five|six|seven|eight|nine"

    parent_path = PosixPath().parent.absolute()
    file_path = parent_path / "SKA" / "inputs" / "aoc01.txt"
    with open(file_path.as_posix(), "r") as file:
        for line in file:
            x = [
                *map(
                    {n: str(i % 9 + 1) for i, n in enumerate(regex.split("|"))}.get,
                    re.findall(rf"(?=({regex}))", line),
                )
            ]

            if len(x) == 0:
                print("no digits found")
            elif len(x) == 1:
                number = x[0] + x[0]
                sum += int(number)
            else:
                number = x[0] + x[-1]
                sum += int(number)

    print(sum)


if __name__ == "__main__":
    main()
